A51-custom-backend-metrics.md

```diff
@@ -0,0 +1,387 @@
+// the default is 30s.
+// In the future, it might be useful to let users specify minimum report interval and default 
+// report interval separately. If the default report interval is not configured or badly configured 
+// non-positive, the default is 1 min. 
+// public OpenRcaServiceImpl(long minInterval, TimeUnit timeUnit, long 
+// defaultInterval, TimeUnit)
+public OrcaServiceImpl(long minReportInterval, TimeUnit unit);
+
+// Creates an OOB service using default minimum report interval = 30s.
+public OrcaServiceImpl();
+
+// Update the metrics value corresponding to the specified key.
+public void setUtilizationMetric(String key, double value);
+
+// Replace the whole metrics data using the specified map.
+public void setAllUtilizationMetrics(Map<String, Double> metrics);
+
+// Remove the metrics data entry corresponding to the specified key.
+public void deleteUtilizationMetric(String key);
+
+// Update the CPU utilization metrics data.
+public void setCPUUtilizationMetric(double value);
+
+// Clear the CPU utilization metrics data.
+public void deleteCPUUtilizationMetric();
+
+// Update the memory utilization metrics data.
+public void setMemoryUtilizationMetric(double value);
+
+// Clear the memory utilization metrics data.
+public void deleteMemoryUtilizationMetric();
+}
+```
+
+At the client side, Java will provide a utility function to wrap the `LoadBalancer.Helper`,
+usable by both WRR policy and xDS LB policies. The `LoadBalancer.Helper` manages the OOB stream
+lifecycle and accepts subscriptions from any routing hierarchy.
+The OOB stream will also gracefully disable itself if the server integration is not complete, 
+so users do not have to worry about client and server deployment coordination.
+
+The following pseudocode shows how a custom load balancing policy instantiates the helper wrapper, 
+installs the out-of-band metrics reporting mechanism, and then consumes the metrics reports, 
+which only requires users to register an `OrcaOobReportListener` that contains their own business logic:
+
+```java
+class CustomLoadBalancer extends LoadBalancer {
+  private final Helper helper;  // the original Helper
+  void handleNewServer(address) {
+  // Create OrcaHelperWrapper and a separate listener for each Subchannel
+  OrcaListener listener = new OrcaOobReportListener();
+  OrcaHelperWrapper orcaWrapper = OrcaUtil.newOrcaHelperWrapper(helper, listener);
+  orcaWrapper.configure(
+      OrcaReportingConfig.newBuilder().setInterval(40, SECONDS).build()); 
+  Subchannel subchannel = orcaWrapper.asHelper().create(
+  CreateSubchannelArgs.newBuilder().setAddresses(address).build());
+  }
+}
+
+// Users implement a listener and pass it to the client library to receive
+// callbacks upon metric report updates from the backend.
+public interface OrcaOobReportListener {
+  //receive load report in the format of ORCA protocol.
+  void onLoadReport(OrcaLoadReport report);
```

Comment:
Looks like we have an issue is the current implementation that you can't distinguish which server this arrived from, yet the same listener is called for an entire LB policy. Seems there will need to be more glue here in some way.

Reply:
Good catch. Yeah i agree that the `OrcaOobReportListener` passed to `OrcaReportingHelper` maps to a single helper. The helper is shared by all the OOB streams, so every server's metrics data is calling the same listener.
What about adding `Subchannel` to become `OrcaOobReportListener.onLoadReport(OrcaLoadReport report, Subchannel subchannel) `, maybe this can apply to per request listener as well.

Reply:
Adding a `Subchannel` argument would certainly be closer, but there are some complications. First and easiest to resolve, we'll need to make sure to pass the correct Subchannel instance; different levels of the LB hierarchy see a subchannel at different wrapping levels. We'll need to remember the proper identity of the [subchannel when adding the listener](https://github.com/grpc/grpc-java/blob/a5829107c3dd771a43cb2ef9c6eb967aa13b8aa3/xds/src/main/java/io/grpc/xds/OrcaOobUtil.java#L278), and the proper identity of the subchannel is only known after the `if (augmented)` block, so we'd need to move that line that adds the listiner.

Second issue is that passing the subchannel through this callback API makes composition more difficult. It only works if the OrcaOobUtil is the top/last wrapper of the subchannel. If the subchannel is wrapped again after OrcaOobUtil, then using the subchannel easily becomes buggy. I'm thinking that it may be better to pass the listener inside CreateSubchannelArgs.

Reply:
> I'm thinking that it may be better to pass the listener inside CreateSubchannelArgs.

oh then it becomes similar to per request case where the user can track which listener is used for each subchannel/oob server. And they will be able to associate multiple listeners to a helper.

Reply:
FWIW, in C-core, the API is to add a separate OOB watcher for each subchannel.

Reply:
@ejona Passing listener via CreateSubchannelArgs helps to solve the problem for leaf LB policy, it allows each subchannel have one listener. But seems not for non-leaf LB policy.

> different levels of the LB hierarchy see a subchannel at different wrapping levels. We'll need to remember the proper identity...

For an non-leaf LB hierarchy that may not use `CreateSubchannelArgs` directly, passing a [listener](https://github.com/grpc/grpc-java/blob/a5829107c3dd771a43cb2ef9c6eb967aa13b8aa3/xds/src/main/java/io/grpc/xds/OrcaOobUtil.java#L113) for each `helper` is still useful, e.g. when they put one callback for each child policy.

If we have a specific  CreateSubchannelArgs key for OOB listener, then we have to remove it after adding it to the listeners set of the OrcaReportingState, so that we do not adding them multiple times.

Reply:
We had a offline discussion that the current java `OrcaOobUtil` implementation needs to be changed, to allow to set multiple listeners to each subchannel and serve the needs of LB policies in different levels of LB hierarchy. We will probably leverage subchannel attributes to store the listeners and add them all at once to the OOB stream at the root of the LB hierarchy. This is because otherwise subchannel can be wrapped subchannel and cause problems. I will reflect the requirement in the gRFC.